sql/analyzer/stored_procedures.go

```diff
@@ -254,10 +251,105 @@ func applyProcedures(ctx *sql.Context, a *Analyzer, n sql.Node, scope *Scope, se
 		return n, transform.SameTree, nil
 	}
 
-	scope, err := loadStoredProcedures(ctx, a, n, scope, sel)
+	call, _, err := transform.Node(n, func(n sql.Node) (sql.Node, transform.TreeIdentity, error) {
+		call, ok := n.(*plan.Call)
+		if !ok {
+			return n, transform.SameTree, nil
+		}
+		if scope.IsEmpty() {
+			scope = scope.withProcedureCache(NewProcedureCache())
+		}
+		if call.AsOf != nil && !scope.enforceReadOnly {
+			scope.enforceReadOnly = true
+			defer func() {
+				scope.enforceReadOnly = false
+			}()
+		}
+
+		esp, err := a.Catalog.ExternalStoredProcedure(ctx, call.Name, len(call.Params))
+		if err != nil {
+			return nil, transform.SameTree, err
+		}
+		if esp != nil {
+			externalProcedure, err := resolveExternalStoredProcedure(ctx, *esp)
+			if err != nil {
+				return nil, transform.SameTree, err
+			}
+			return call.WithProcedure(externalProcedure), transform.NewTree, nil
+		}
+
+		if spdb, ok := call.Database().(sql.StoredProcedureDatabase); ok {
+			procedure, ok, err := spdb.GetStoredProcedure(ctx, call.Name)
+			if err != nil {
+				return nil, transform.SameTree, err
+			}
+			if !ok {
+				err := sql.ErrStoredProcedureDoesNotExist.New(call.Name)
+				if call.Database().Name() == "" {
+					return nil, transform.SameTree, fmt.Errorf("%w; this might be because no database is selected", err)
+				}
+				return nil, transform.SameTree, err
+			}
+			parsedProcedure, err := parse.Parse(ctx, procedure.CreateStatement)
+			if err != nil {
+				return nil, transform.SameTree, err
+			}
+			parsedProcedure, _, err = transform.Node(parsedProcedure, func(n sql.Node) (sql.Node, transform.TreeIdentity, error) {
+				switch n := n.(type) {
+				case *plan.Call:
+					newCall := n
+					tree := transform.SameTree
+					if newCall.Database() == nil || newCall.Database().Name() == "" {
+						newNode, err := newCall.WithDatabase(spdb)
+						if err != nil {
+							return nil, transform.SameTree, err
+						}
+						newCall = newNode.(*plan.Call)
+						tree = transform.NewTree
+					}
+					if newCall.AsOf == nil {
+						newCall = newCall.WithAsOf(call.AsOf)
+						tree = transform.NewTree
+					}
+					if tree == transform.NewTree {
+						return newCall, transform.NewTree, nil
+					}
+				case *plan.ShowCreateTable:
+					if n.AsOf == nil {
+						return n.WithAsOf(call.AsOf), transform.NewTree, nil
+					}
```

Comment:
(just a thought) Not necessary for this change, but just a thought to consider in the future... it might be convenient to have an interface for nodes that support `AsOf` so we don't have to special case them in place like this and could just grab them all with a case on the interface. Probably overkill just for this one method, but might be something to consider if we see this pattern with `AsOf` in other places.

Reply:
I definitely was going to go through the trouble of adding an interface, but when I saw that there were only 4 cases I decided to just write them here. One benefit from the interface approach is that it signals to future developers that may base their code off of existing plans that they also need to implement the interface. Having bespoke logic here means that the chances of another developer adding an entry here will only be because they've observed incorrect behavior. So especially after typing this all here, I think it's definitely beneficial to add the interface.